lib/stdlib/src/gen_statem.erl

```diff
@@ -0,0 +1,1095 @@
+      | {Name :: atom(), Node :: atom()}
+      | pid().
+-type debug_opt() ::
+	{'debug',
+	 Dbgs ::
+	   ['trace' | 'log' | 'statistics' | 'debug'
+	    | {'logfile', string()}]}.
+-type start_opt() ::
+	debug_opt()
+      | {'timeout', Time :: timeout()}
+      | {'spawn_opt', SOpts :: [proc_lib:spawn_option()]}.
+-type start_ret() ::  {'ok', pid()} | 'ignore' | {'error', term()}.
+
+
+
+%% Start a state machine
+-spec start(
+	Module :: module(), Args :: term(), Options :: [start_opt()]) ->
+		   start_ret().
+start(Module, Args, Options) ->
+    gen:start(?MODULE, nolink, Module, Args, Options).
+%%
+-spec start(
+	ServerName :: server_name(),
+	Module :: module(), Args :: term(), Options :: [start_opt()]) ->
+		   start_ret().
+start(ServerName, Module, Args, Options) ->
+    gen:start(?MODULE, nolink, ServerName, Module, Args, Options).
+
+%% Start and link to a state machine
+-spec start_link(
+	Module :: module(), Args :: term(), Options :: [start_opt()]) ->
+		   start_ret().
+start_link(Module, Args, Options) ->
+    gen:start(?MODULE, link, Module, Args, Options).
+%%
+-spec start_link(
+	ServerName :: server_name(),
+	Module :: module(), Args :: term(), Options :: [start_opt()]) ->
+		   start_ret().
+start_link(ServerName, Module, Args, Options) ->
+    gen:start(?MODULE, link, ServerName, Module, Args, Options).
+
+%% Stop a state machine
+-spec stop(ServerRef :: server_ref()) -> ok.
+stop(ServerRef) ->
+    gen:stop(ServerRef).
+%%
+-spec stop(
+	ServerRef :: server_ref(),
+	Reason :: term(),
+	Timeout :: timeout()) -> ok.
+stop(ServerRef, Reason, Timeout) ->
+    gen:stop(ServerRef, Reason, Timeout).
+
+%% Send an event to a state machine that arrives with type 'event'
+-spec cast(ServerRef :: server_ref(), Msg :: term()) -> ok.
+cast({global,Name}, Msg) ->
+    try	global:send(Name, cast(Msg)) of
+	_ -> ok
+    catch
+	_:_ -> ok
+    end;
+cast({via,RegMod,Name}, Msg) ->
+    try	RegMod:send(Name, cast(Msg)) of
+	_ -> ok
+    catch
+	_:_ -> ok
+    end;
+cast({Name,Node} = ServerRef, Msg) when is_atom(Name), is_atom(Node) ->
+    do_send(ServerRef, cast(Msg));
+cast(ServerRef, Msg) when is_atom(ServerRef) ->
+    do_send(ServerRef, cast(Msg));
+cast(ServerRef, Msg) when is_pid(ServerRef) ->
+    do_send(ServerRef, cast(Msg)).
+
+%% Call a state machine (synchronous; a reply is expected) that
+%% arrives with type {call,Client}
+-spec call(ServerRef :: server_ref(), Request :: term()) -> Reply :: term().
+call(ServerRef, Request) ->
+    call(ServerRef, Request, infinity).
+%%
+-spec call(
+	ServerRef :: server_ref(),
+	Request :: term(),
+	Timeout :: timeout()) ->
+		  Reply :: term().
+call(ServerRef, Request, infinity) ->
+    try gen:call(ServerRef, '$gen_call', Request, infinity) of
+	{ok,Reply} ->
+	    Reply
+    catch
+	Class:Reason ->
+	    erlang:raise(
+	      Class,
+	      {Reason,{?MODULE,call,[ServerRef,Request,infinity]}},
+	      erlang:get_stacktrace())
+    end;
+call(ServerRef, Request, Timeout) ->
```

Comment:
The proxy code comes on lines 298..332 so that diff is somewhat misleading.

The reason is that a synchronous call with timeout < infinity is a wart that causes problems so if you use it you might as well take the performance hit of getting it right.  We hesitate to do this in e.g `gen_server` for backwards compatibility reasons, but why not do it while writing a new behaviour.

You should use call with infinite timeout and handle timeouts on the server side.  That is a better strategy.  Also note that I took the liberty to change the default timeout to `infinity` instead of `gen_server`'s arbitrary well known 5000 ms.  Why not do it right in a new behaviour?